tools/xkbcli-scaffold-new-layout.py

```diff
@@ -0,0 +1,323 @@
+#  Software is furnished to do so, subject to the following conditions:
+#
+#  The above copyright notice and this permission notice (including the next
+#  paragraph) shall be included in all copies or substantial portions of the
+#  Software.
+#
+#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+#  DEALINGS IN THE SOFTWARE.
+
+
+# This script creates the necessary scaffolding to create a custom keyboard
+# layout or option. It does not actually configure anything, it merely creates
+# the required directory structure and file scaffolding for the key
+# configurations to be added by the user.
+#
+
+import argparse
+import logging
+import os
+import re
+import sys
+
+from pathlib import Path
+from textwrap import dedent
+
+
+# Default values are set by meson but to make this usable directly within the
+# git source tree, use some sensible default values and return those where the
+# meson define hasn't been replaced.
+def default_value(key):
+    defaults = {
+        'extrapath': ('@XKBEXTRAPATH@', '/etc/xkb'),
+        'rules': ('@DEFAULT_XKB_RULES@', 'evdev'),
+    }
+
+    mesondefault, default = defaults[key]
+    if mesondefault.startswith('@') and mesondefault.endswith('@'):
+        return default
+    else:
+        return mesondefault
+
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger('xkbcli')
+logger.setLevel(logging.INFO)
+
+
+def create_directory_structure(basedir):
+    basedir.mkdir(exist_ok=True)
+    # Note: we skip geometry
+    for d in ['compat', 'keycodes', 'rules', 'symbols', 'types']:
+        (basedir / d).mkdir(exist_ok=True)
+
+
+def create_rules_template(base_path, ruleset, layout, option):
+    rules = Path(base_path) / 'rules' / ruleset
+    if rules.exists():
+        logger.warning(f'Rules file {rules} already exists, skipping')
+        return
+
+    with open(rules, 'w') as fd:
+        header = dedent(f'''\
+            // generated by xkbcli scaffold-new-layout
+
+            // Note: no rules file entries are required for for a custom layout
+
+            ''')
+        fd.write(header)
+
+        if option:
+            group, section = option
+            option_template = dedent(f'''\
+            // This section maps XKB option "{group}:{section}" to the '{section}' section in the
+            // 'symbols/{group}' file.
+            //
+            ! option = symbols
+              {group}:{section} = +{group}({section})
+
+            ''')
+            fd.write(option_template)
+
+        footer = dedent(f'''\
+            // Include the system '{ruleset}' file
```

Comment:
I wonder if `--user` is used we should also somehow hook up `/etc/xkb` here? Might be a little difficult as it might (probably) not exist...

Reply:
yeah, and we'd need to run as root. And it's a niche case anyway, so overall I think it's not worth the bother. Or at least not until we see that users actually need this, right now it's all a bit of stab in the dark.